File: app/src/main/java/com/zpf/myplan/ui/widget/navigation/BottomNavigationMenuView.java
package com.zpf.myplan.ui.widget.navigation;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;

import com.zpf.myplan.R;

public class BottomNavigationMenuView extends ViewGroup {
    private static final String TAG = "BottomNavigationMenuView";

    public BottomNavigationMenuView(Context context) {
        super(context);
    }

    //    private MenuInflater menuInflater;
    private BottomNavigationMenuInflater menuInflater;
    private BottomNavigationMenu menu;
    private BottomMenuItemView chooseItemView;

    public BottomNavigationMenuView(Context context, AttributeSet attrs) {
        super(context, attrs);
        this.menu = new BottomNavigationMenu();
        menuInflater = new BottomNavigationMenuInflater(getContext());
        @SuppressLint("Recycle") TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.BottomNavigationMenuView);
        if (typedArray.hasValue(R.styleable.BottomNavigationMenuView_menu)) {
            inflateMenu(typedArray.getResourceId(R.styleable.BottomNavigationMenuView_menu, 0));
        }

        for (int i = 0; i < menu.getItems().size(); i++) {
            BottomMenuItem bottomMenuItem = menu.getItems().get(i);
            BottomMenuItemView itemView = new BottomMenuItemView(context, bottomMenuItem.getTitle(), bottomMenuItem.getIcon());
            addView(itemView);
            itemView.setOnClickListener(v -> {
                BottomMenuItemView thisView = (BottomMenuItemView) v;
                thisView.onChoose();
                if (chooseItemView!=null){
                    chooseItemView.onUnChoose();
                }
                chooseItemView = thisView;
                Log.d(TAG, "BottomNavigationMenuView: ");
            });
        }
    }

    public void inflateMenu(int resId) {
        menuInflater.inflate(resId, menu);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        if (getChildCount() > 0) {
            int count = getChildCount();
            for (int i = 0; i < count; i++) {
                View child = getChildAt(i);
                measureChild(child, widthMeasureSpec, heightMeasureSpec);
            }
        }
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        int count = getChildCount();
        int parentWith = ((ViewGroup) getParent()).getWidth();
        int parentHeight = ((ViewGroup) getParent()).getHeight();
        int itemWith = parentWith / count;
        int widget = parentHeight - 10;
        for (int i = 0; i < count; i++) {
            View view = getChildAt(i);
            int height = view.getMeasuredHeight();
            int width = view.getMeasuredWidth();
            int left = itemWith * i + (itemWith - width) / 2;
            view.layout(left, widget - height, left + width, widget);
        }
    }
}

File: app/src/main/java/com/zpf/myplan/ui/widget/navigation/BottomNavigationMenuView.kt
package com.zpf.myplan.ui.widget.navigation

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.view.View
import android.view.ViewGroup
import com.zpf.myplan.R
import com.zpf.myplan.navigation.NavigationManager
import com.zpf.myplan.ui.event.EventFragment
import com.zpf.myplan.ui.mine.MineFragment
import com.zpf.myplan.ui.note.NoteFragment
import com.zpf.myplan.ui.plan.PlanFragment
import com.zpf.myplan.ui.time.TimeFragment

class BottomNavigationMenuView : ViewGroup {
    constructor(context: Context?) : super(context) {}

    //    private MenuInflater menuInflater;
    private var mNavigationManager: NavigationManager? = null
    private var menuInflater: BottomNavigationMenuInflater? = null
    private var menu: BottomNavigationMenu? = null
    private var chooseItemView: BottomMenuItemView? = null

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        menu = BottomNavigationMenu()
        menuInflater = BottomNavigationMenuInflater(getContext())
        @SuppressLint("Recycle") val typedArray =
            context.obtainStyledAttributes(attrs, R.styleable.BottomNavigationMenuView)
        if (typedArray.hasValue(R.styleable.BottomNavigationMenuView_menu)) {
            inflateMenu(typedArray.getResourceId(R.styleable.BottomNavigationMenuView_menu, 0))
        }
        for (i in menu!!.items.indices) {
            val bottomMenuItem = menu!!.items[i]
            val itemView = BottomMenuItemView(context, bottomMenuItem.title, bottomMenuItem.icon, i)
            addView(itemView)
            itemView.setOnClickListener { v: View ->
                val thisView = v as BottomMenuItemView
                thisView.onChoose()
                if (chooseItemView != null) {
                    chooseItemView!!.onUnChoose()
                }
                chooseItemView = thisView
                switchFragment(itemView.id)
                Log.d(
                    TAG,
                    "BottomNavigationMenuView: "
                )
            }
            if (i == 0) {
                chooseItemView = itemView
                itemView.onChoose()
            }
        }
    }

    private fun inflateMenu(resId: Int) {
        menuInflater!!.inflate(resId, menu)
    }

    fun setNavigationManager(navigationManager: NavigationManager) {
        mNavigationManager = navigationManager
        mNavigationManager?.switchFragment<PlanFragment>()
    }

    fun switchFragment(id: Int) {
        when (id) {
            0 -> mNavigationManager?.switchFragment<PlanFragment>()
            1 -> mNavigationManager?.switchFragment<NoteFragment>()
            2 -> mNavigationManager?.switchFragment<EventFragment>()
            3 -> mNavigationManager?.switchFragment<TimeFragment>()
            4 -> mNavigationManager?.switchFragment<MineFragment>()
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        if (childCount > 0) {
            val count = childCount
            for (i in 0 until count) {
                val child = getChildAt(i)
                measureChild(child, widthMeasureSpec, heightMeasureSpec)
            }
        }
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        val count = childCount
        val parentWith = (parent as ViewGroup).width
        val parentHeight = (parent as ViewGroup).height
        val itemWith = parentWith / count
        val widget = parentHeight - 10
        for (i in 0 until count) {
            val view = getChildAt(i)
            val height = view.measuredHeight
            val width = view.measuredWidth
            val left = itemWith * i + (itemWith - width) / 2
            view.layout(left, widget - height, left + width, widget)
        }
    }

    companion object {
        private const val TAG = "BottomNavigationMenuView"
    }
}